skills/contextual-embeddings/guide.ipynb

```diff
@@ -0,0 +1,1372 @@
+    "\n",
+    "In traditional RAG, documents are typically split into smaller chunks for efficient retrieval. While this approach works well for many applications, it can lead to problems when individual chunks lack sufficient context. Contextual Embeddings solve this problem by adding relevant context to each chunk before embedding. This method improves the quality of each embedded chunk, allowing for more accurate retrieval and thus better overall performance. Averaged across all data sources we tested, Contextual Embeddings reduced the top-20-chunk retrieval failure rate by 35%.\n",
+    "\n",
+    "The same chunk-specific context can also be used with BM25 search to further improve retrieval performance. We introduce this technique in the “Contextual BM25” section.\n",
+    "\n",
+    "In this guide, we'll demonstrate how to build and optimize a Contextual Retrieval system using a dataset of 9 codebases as our knowledge base. We'll walk through:\n",
+    "\n",
+    "1) Setting up a basic retrieval pipeline to establish a baseline for performance.\n",
+    "\n",
+    "2) Contextual Embeddings: what it is, why it works, and how prompt caching makes it practical for production use cases.\n",
+    "\n",
+    "3) Implementing Contextual Embeddings and demonstrating performance improvements.\n",
+    "\n",
+    "4) Contextual BM25: improving performance with *contextual* BM25 hybrid search.\n",
+    "\n",
+    "5) Improving performance with reranking,\n",
+    "\n",
+    "### Evaluation Metrics & Dataset:\n",
+    "\n",
+    "We use a pre-chunked dataset of 9 codebases - all of which have been chunked according to a basic character splitting mechanism. Our evaluation dataset contains 248 queries - each of which contains a 'golden chunk.' We'll use a metric called Pass@k to evaluate performance. Pass@k checks whether or not the 'golden document' was present in the first k documents retrieved for each query. Contextual Embeddings in this case helped us to improve Pass@10 performance from ~87% --> ~95%.\n",
+    "\n",
+    "You can find the code files and their chunks in `data/codebase_chunks.json` and the evaluation dataset in `data/evaluation_set.jsonl`\n",
+    "\n",
+    "#### Additional Notes:\n",
+    "\n",
+    "Prompt caching is helpful in managing costs when using this retrieval method. This feature is currently available on Anthropic's 1P API, and is coming soon to our 3P partner environments in AWS Bedrock and GCP Vertex. We know that many of our customers leverage AWS Knowledge Bases and GCP Vertex AI APIs when building RAG solutions, and this method can be used on either platform with a bit of customization. Consider reaching out to Anthropic or your AWS/GCP account team for guidance on this!\n",
```

Comment:
We should set a reminder to update this when launches on Bedrock! Also, maybe a link to prompt caching documentation would be helpful here

Reply:
yes we will for sure do this!